File: testSuite/scripts/decayingDarkMatterRetention.pl
#!/usr/bin/env perl
use strict;
use warnings;
use PDL;
use PDL::NiceSlice;
use PDL::IO::HDF5;
use PDL::GSL::INTEG;
use PDL::Constants qw(PI);

# Compute the fraction of particles, and kick energy retained in a halo for decaying dark matter models. Uses a Monte Carlo
# simulation to evaluate these fractions, which is used as a target dataset to test our numerical calculations.
# Andrew Benson (01-May-2024)

# Specify the velocity dispersion and the escape velocity.
my $velocityDispersion = pdl 20.0;
my $velocityEscape     = pdl 60.0;

# Generate a sequence of velocity kicks for which to compute the retained fractions.
my $velocityKicks      = pdl sequence(130)+1.0;

# Generate a sequence of escape velocities for which to compute the retained fractions.
my $velocityEscapes    = pdl sequence( 10)+60.0;

# Specify the number of Monte Carlo trials.
my $countTrials = 30;

# Specify the number of particles to sample for each trial.
my $countSample = 1000000;

# Create arrays to store results.
my $fractionsRetained = pdl zeroes($countTrials,nelem($velocityKicks),nelem($velocityEscapes));
my $energiesRetained  = pdl zeroes($countTrials,nelem($velocityKicks),nelem($velocityEscapes));

# Iterate over escape velocities.
my $velocityWidth = $velocityDispersion->copy();
for(my $j=0;$j<nelem($velocityEscapes);++$j) {
    my $velocityEscape   = $velocityEscapes->(($j));
    print "Velocity escape = ".$velocityEscape." [".($j+1)." of ".nelem($velocityEscapes)."]\n";
    # Compute the width of the truncated Maxwell-Boltzmann distribution (truncated at the escape velocity) required to give a mean
    # squared velocity of 3 times the velocity dispersion. We use a simple iterative scheme here.
    my $velocitySquaredMeanTarget = 3.0*$velocityDispersion**2;
    my $boostFactor               = pdl 1.0;
    my $velocityMeanSquared;
    for(my $i=0;$i<10;++$i) {
	$velocityWidth         .= $boostFactor*$velocityDispersion;
	my $limit               = 1000     ;
	my $key                 =    5     ;
	my $toleranceRelative   =    1.0e-4;
	my $toleranceAbsolute   =    0.0e+0;
	($velocityMeanSquared)  = gslinteg_qag(\&velocityMeanSquaredIntegrand ,0.0,$velocityEscape,$toleranceRelative,$toleranceAbsolute,$limit,$key);
	my ($normalization   )  = gslinteg_qag(\&velocitNormalizationIntegrand,0.0,$velocityEscape,$toleranceRelative,$toleranceAbsolute,$limit,$key);
	$velocityMeanSquared   /= $normalization;
	$boostFactor           *= sqrt($velocitySquaredMeanTarget/$velocityMeanSquared);

    }
    print "\tComputed boost factor of: ".$boostFactor."\n";
    print "\t\tMean squared velocity is ".$velocityMeanSquared." (target was ".$velocitySquaredMeanTarget.")\n";
    # Construct the truncated Maxwell-Boltzmann distribution and find the cumulative distribution function.
    my $countVelocity = 100000;
    my $velocity      = pdl sequence($countVelocity)/($countVelocity-1)*$velocityEscape;
    my $pdf           = $velocity**2*exp(-0.5*($velocity/$velocityWidth)**2);
    my $cdf           = $pdf->cumusumover()/$pdf->sum();
    # Iterate over velocity kicks.
    for(my $i=0;$i<nelem($velocityKicks);++$i) {
	# Iterate over trials.
	my $velocityKick   = $velocityKicks->(($i));
	print "\tVelocity kick = ".$velocityKick." [".($i+1)." of ".nelem($velocityKicks)."]\n";
	for(my $trial=0;$trial<$countTrials;++$trial) {
	    # Generate random velocities and kick angles.
	    my $x1 = pdl random($countSample);
	    my $x2 = pdl random($countSample);
	    (my $velocities) = interpolate($x1,$cdf,$velocity);
	    my $cosTheta     = 2.0*$x2-1.0;
	    # Compute the final velocity.
	    my $velocityFinal = sqrt($velocities**2+$velocityKick**2+2.0*$velocities*$velocityKick*$cosTheta);
	    # Find those particles that are retained.
	    my $retained = which($velocityFinal < $velocityEscape);
	    # Compute the energy gain of retain particles.
	    my $energyGain = sum(0.5*$velocityFinal->($retained)**2-0.5*$velocities->($retained)**2)/$countSample;
	    # Store the retained fractions.
	    $energiesRetained ->(($trial),($i),($j)) .= $energyGain/(0.5*$velocityKick**2);
	    $fractionsRetained->(($trial),($i),($j)) .= nelem($retained)/$countSample;
	}
    }
}

# Compute the mean fractions and their uncertainties.
my $fractionRetained            = $fractionsRetained->average();
my $fractionRetainedUncertainty = sqrt(($fractionsRetained->pow(2)->average()-$fractionRetained->pow(2))/$countTrials);
my $energyRetained              = $energiesRetained ->average();
my $energyRetainedUncertainty   = sqrt(($energiesRetained ->pow(2)->average()-$energyRetained  ->pow(2))/$countTrials);

# Store results to file.
my $output = new PDL::IO::HDF5(">testSuite/data/decayingDarkMatterRetention.hdf5");
$output->dataset('velocityKick'               )->set($velocityKicks              );
$output->dataset('velocityEscape'             )->set($velocityEscapes             );
$output->dataset('fractionRetained'           )->set($fractionRetained           );
$output->dataset('fractionRetainedUncertainty')->set($fractionRetainedUncertainty);
$output->dataset('energyRetained'             )->set($energyRetained             );
$output->dataset('energyRetainedUncertainty'  )->set($energyRetainedUncertainty  );
$output->attrSet(velocityDispersion => $velocityDispersion);

exit;

sub velocityMeanSquaredIntegrand {
    # Velocity distribution function.
    my ($velocity) = @_;
    return sqrt(2.0/PI)*$velocityWidth*($velocity/$velocityWidth)**4*exp(-0.5*($velocity/$velocityWidth)**2);
}

sub velocitNormalizationIntegrand {
    # Velocity distribution function.
    my ($velocity) = @_;
    return sqrt(2.0/PI)*($velocity/$velocityWidth)**2/$velocityWidth*exp(-0.5*($velocity/$velocityWidth)**2);
}

File: testSuite/scripts/decayingDarkMatterRetention.py
#!/usr/bin/env python3
import numpy as np
import h5py
from scipy import integrate

# Compute the fraction of particles, and kick energy retained in a halo for decaying dark matter models. Uses a Monte Carlo
# simulation to evaluate these fractions, which is used as a target dataset to test our numerical calculations.
# Andrew Benson (01-May-2024; ported to Python)

# Specify the velocity dispersion and the escape velocity.
velocity_dispersion = 20.0
velocity_escape     = 60.0

# Generate a sequence of velocity kicks for which to compute the retained fractions.
velocity_kicks   = np.arange(1, 131, dtype=float)  # 1..130

# Generate a sequence of escape velocities for which to compute the retained fractions.
velocity_escapes = np.arange(60, 70,  dtype=float)  # 60..69

# Specify the number of Monte Carlo trials.
count_trials = 30

# Specify the number of particles to sample for each trial.
count_sample = 1_000_000

# Create arrays to store results.
fractions_retained = np.zeros((count_trials, len(velocity_kicks), len(velocity_escapes)))
energies_retained  = np.zeros((count_trials, len(velocity_kicks), len(velocity_escapes)))

# Iterate over escape velocities.
velocity_width = velocity_dispersion
for j, v_escape in enumerate(velocity_escapes):
    print(f"Velocity escape = {v_escape} [{j+1} of {len(velocity_escapes)}]")

    # Compute the width of the truncated Maxwell-Boltzmann distribution (truncated at the escape
    # velocity) required to give a mean squared velocity of 3 times the velocity dispersion. We
    # use a simple iterative scheme here.
    velocity_squared_mean_target = 3.0 * velocity_dispersion**2
    boost_factor = 1.0
    velocity_mean_squared = None
    for _ in range(10):
        velocity_width = boost_factor * velocity_dispersion

        def velocity_mean_squared_integrand(v):
            return (np.sqrt(2.0 / np.pi) * velocity_width
                    * (v / velocity_width)**4
                    * np.exp(-0.5 * (v / velocity_width)**2))

        def velocity_normalization_integrand(v):
            return (np.sqrt(2.0 / np.pi)
                    * (v / velocity_width)**2 / velocity_width
                    * np.exp(-0.5 * (v / velocity_width)**2))

        velocity_mean_squared, _ = integrate.quad(
            velocity_mean_squared_integrand, 0.0, v_escape,
            limit=1000, epsrel=1.0e-4, epsabs=0.0
        )
        normalization, _ = integrate.quad(
            velocity_normalization_integrand, 0.0, v_escape,
            limit=1000, epsrel=1.0e-4, epsabs=0.0
        )
        velocity_mean_squared /= normalization
        boost_factor *= np.sqrt(velocity_squared_mean_target / velocity_mean_squared)

    print(f"\tComputed boost factor of: {boost_factor}")
    print(f"\t\tMean squared velocity is {velocity_mean_squared} (target was {velocity_squared_mean_target})")

    # Construct the truncated Maxwell-Boltzmann distribution and find the cumulative distribution function.
    count_velocity = 100_000
    velocity       = np.arange(count_velocity) / (count_velocity - 1) * v_escape
    pdf            = velocity**2 * np.exp(-0.5 * (velocity / velocity_width)**2)
    cdf            = np.cumsum(pdf) / pdf.sum()

    # Iterate over velocity kicks.
    for i, v_kick in enumerate(velocity_kicks):
        print(f"\tVelocity kick = {v_kick} [{i+1} of {len(velocity_kicks)}]")
        # Iterate over trials.
        for trial in range(count_trials):
            # Generate random velocities and kick angles.
            x1 = np.random.uniform(size=count_sample)
            x2 = np.random.uniform(size=count_sample)
            velocities = np.interp(x1, cdf, velocity)
            cos_theta  = 2.0 * x2 - 1.0
            # Compute the final velocity.
            velocity_final = np.sqrt(velocities**2 + v_kick**2 + 2.0 * velocities * v_kick * cos_theta)
            # Find those particles that are retained.
            retained = velocity_final < v_escape
            # Compute the energy gain of retained particles.
            energy_gain = np.sum(
                0.5 * velocity_final[retained]**2 - 0.5 * velocities[retained]**2
            ) / count_sample
            # Store the retained fractions.
            energies_retained [trial, i, j] = energy_gain / (0.5 * v_kick**2)
            fractions_retained[trial, i, j] = retained.sum() / count_sample

# Compute the mean fractions and their uncertainties.
fraction_retained             = fractions_retained.mean(axis=0)
fraction_retained_uncertainty = np.sqrt(
    ((fractions_retained**2).mean(axis=0) - fraction_retained**2) / count_trials
)
energy_retained               = energies_retained.mean(axis=0)
energy_retained_uncertainty   = np.sqrt(
    ((energies_retained**2).mean(axis=0) - energy_retained**2) / count_trials
)

# Store results to file.
with h5py.File("testSuite/data/decayingDarkMatterRetention.hdf5", "w") as f:
    f.create_dataset("velocityKick",                data=velocity_kicks)
    f.create_dataset("velocityEscape",              data=velocity_escapes)
    f.create_dataset("fractionRetained",            data=fraction_retained)
    f.create_dataset("fractionRetainedUncertainty", data=fraction_retained_uncertainty)
    f.create_dataset("energyRetained",              data=energy_retained)
    f.create_dataset("energyRetainedUncertainty",   data=energy_retained_uncertainty)
    f.attrs["velocityDispersion"] = velocity_dispersion
